x/pocketcore/handler.go

```diff
@@ -1,62 +1,126 @@
 package pocketcore
 
 import (
+	"encoding/hex"
 	"fmt"
 	"github.com/pokt-network/pocket-core/x/pocketcore/keeper"
 	"github.com/pokt-network/pocket-core/x/pocketcore/types"
+	"github.com/pokt-network/posmint/crypto"
 	sdk "github.com/pokt-network/posmint/types"
 )
 
 // NewHandler returns a handler for "pocketCore" type messages.
 func NewHandler(keeper keeper.Keeper) sdk.Handler {
 	return func(ctx sdk.Context, msg sdk.Msg) sdk.Result {
 		switch msg := msg.(type) {
-		case types.MsgProofOfRelays:
-			return handleProofBatchMessage(ctx, keeper, msg)
+		case types.MsgProof:
+			return handleProofMsg(ctx, keeper, msg)
+		case types.MsgClaimProof:
+			return handleClaimProofMsg(ctx, keeper, msg)
 		default:
 			errMsg := fmt.Sprintf("Unrecognized pocketcore Msg type: %v", msg.Type())
 			return sdk.ErrUnknownRequest(errMsg).Result()
 		}
 	}
 }
 
-// Handle a message to set name
-func handleProofBatchMessage(ctx sdk.Context, keeper keeper.Keeper, msg types.MsgProofOfRelays) sdk.Result {
-	if keeper.IsPocketSupportedBlockchain(ctx.WithBlockHeight(msg.SessionBlockHeight), msg.Chain) {
-		return types.NewChainNotSupportedErr(types.ModuleName).Result()
+func handleProofMsg(ctx sdk.Context, keeper keeper.Keeper, msg types.MsgProof) sdk.Result {
+	// validate the proof message
+	if err := validateProofMsg(ctx, keeper, msg); err != nil {
+		return err.Result()
 	}
+	// set the unverified proof in the world state
+	keeper.SetUnverifiedProof(ctx, msg)
+	// create the event
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeUnverifiedProof,
+			sdk.NewAttribute(types.AttributeKeyValidator, msg.FromAddress.String()),
+		),
+	})
+	return sdk.Result{Events: ctx.EventManager().Events()}
+}
+
+func handleClaimProofMsg(ctx sdk.Context, keeper keeper.Keeper, msg types.MsgClaimProof) sdk.Result {
+	// validate the claim proof
+	addr, totalRelays, err := validateClaimProofMsg(ctx, keeper, msg)
+	if err != nil {
+		return err.Result()
+	}
+	// valid claim so award coins for relays
+	keeper.AwardCoinsForRelays(ctx, totalRelays, addr)
+	// create the event
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.EventTypeClaimedProof,
+			sdk.NewAttribute(types.AttributeKeyValidator, addr.String()),
+		),
+	})
+	return sdk.Result{Events: ctx.EventManager().Events()}
+}
+
+func validateProofMsg(ctx sdk.Context, keeper keeper.Keeper, msg types.MsgProof) sdk.Error {
+	// if is not a pocket supported blockchain then return not supported error
+	if !keeper.IsPocketSupportedBlockchain(ctx.WithBlockHeight(msg.SessionBlockHeight), msg.Chain) {
```

Comment:
What happens if the proof is from a chain that was supported let's say 3 sessions ago and now it's not supported ( assuming something like this can happen in the future)